test/xbgpu/test_engine.py

```diff
@@ -1452,3 +1490,125 @@ def get_stream_status(stream_name: str) -> bool:
+        # Much like the XBEngine receiver, the BPipeline send-stream transmits
+        # chunks that contain `HEAPS_PER_FENGINE_PER_CHUNK`. Consequently, the
+        # `?capture-stop` request must be issued at least one send-chunk before
+        # data transmission is over to ensure some strength in the test.
+        capture_stop_chunk_index = 10
+        assert capture_stop_chunk_index < n_batches // HEAPS_PER_FENGINE_PER_CHUNK
+        self._patch_send_chunk(monkeypatch, capture_stop_chunk_index, client, requests=capture_stop_requests)
+        timestamp_step = n_samples_between_spectra * n_spectra_per_heap
+
+        # NOTE: The `beam_results` contain data for each heap (or batch) transmitted.
+        # As a result, we multiply the chunk ID at which `?capture-stop` was issued
+        # by `HEAPS_PER_FENGINE_PER_CHUNK` to get the specific heap ID at which data
+        # ceased transmitting for a stopped stream.
+        beam_capture_stop_heap_index = capture_stop_chunk_index * HEAPS_PER_FENGINE_PER_CHUNK
+        corrprod_results, beam_results, _, _ = await self._send_data(
+            mock_recv_streams,
+            mock_send_stream,
+            corrprod_outputs,
+            beam_outputs,
+            heap_factory=lambda batch_index, present: self._default_heap_factory(
+                timestamp_step=timestamp_step,
+                n_ants=n_ants,
+                n_channels_per_substream=n_channels_per_substream,
+                n_spectra_per_heap=n_spectra_per_heap,
+                frequency=frequency,
+                data_value=10,
+                batch_index=batch_index,
+                present=present,
+            ),
+            present=np.ones((n_batches, n_ants), bool),
+            timestamp_step=timestamp_step,
+            n_channels_per_substream=n_channels_per_substream,
+            frequency=frequency,
+            n_spectra_per_heap=n_spectra_per_heap,
+            beam_capture_stop_heap_index=beam_capture_stop_heap_index,
+        )
+
+        # NOTE: During receiving beam data, only data up to `beam_capture_stop_index`
+        # is stored. It is also verified that there is no more data in the stream.
+        # This is largely a sanity check that there is non-zero data for heaps that
+        # were transmitted before the `?capture-stop` request.
+        for beam_output in beam_outputs:
+            np.testing.assert_equal(beam_results[beam_output.name][:beam_capture_stop_heap_index] != 0, True)
+
+        for corrprod_output in corrprod_outputs:
+            # NOTE: The X-engine output is entirely real (no imag component).
+            # The results buffer has shape (n_accumulations,
+            # n_channels_per_substream, n_baselines, COMPLEX). As a result, the
+            # final dimension only has data populated for the 'real' (first) index.
+            np.testing.assert_equal(corrprod_results[corrprod_output.name][..., 0] != 0, True)
```

Comment:
Although looking at this again, I suppose there is scope for me to roll the `corrprod_results` and `beam_results` back to simply numpy `ndarray`s.

Let me know what you think, happy to do so (but will quietly grumble from time to time).